packages/auth/test/integration/flows/totp.test.ts

```diff
@@ -0,0 +1,166 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect, use } from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+import sinonChai from 'sinon-chai';
+import {
+  Auth,
+  multiFactor,
+  signInWithEmailAndPassword,
+  getMultiFactorResolver
+} from '@firebase/auth';
+import { FirebaseError } from '@firebase/app';
+import {
+  cleanUpTestInstance,
+  getTestInstance,
+  getTotpCode,
+  delay,
+  email
+} from '../../helpers/integration/helpers';
+
+import {
+  TotpMultiFactorGenerator,
+  TotpSecret
+} from '../../../src/mfa/assertions/totp';
+
+use(chaiAsPromised);
+use(sinonChai);
+
+describe(' Integration tests: Mfa TOTP', () => {
+  let auth: Auth;
+  let totpSecret: TotpSecret;
+  let displayName: string;
+  beforeEach(async () => {
+    auth = getTestInstance();
+    displayName = 'totp-integration-test';
+  });
+
+  afterEach(async () => {
+    await cleanUpTestInstance(auth, 'totp');
+  });
+
+  it('should not enroll if incorrect totp supplied', async () => {
+    const cr = await signInWithEmailAndPassword(auth, email, 'password');
+    const mfaUser = multiFactor(cr.user);
+    const session = await mfaUser.getSession();
+    totpSecret = await TotpMultiFactorGenerator.generateSecret(session);
+    const totpVerificationCode = getTotpCode(
+      totpSecret.secretKey,
+      totpSecret.codeIntervalSeconds,
+      totpSecret.codeLength
+    );
+
+    const multiFactorAssertion =
+      TotpMultiFactorGenerator.assertionForEnrollment(
+        totpSecret,
+        totpVerificationCode + '0'
+      );
+
+    await expect(
+      mfaUser.enroll(multiFactorAssertion, displayName)
+    ).to.be.rejectedWith('auth/invalid-verification-code');
+  });
+  it('should enroll using correct otp', async () => {
+    const cr = await signInWithEmailAndPassword(auth, email, 'password');
+
+    const mfaUser = multiFactor(cr.user);
+
+    const session = await mfaUser.getSession();
+
+    totpSecret = await TotpMultiFactorGenerator.generateSecret(session);
+
+    const totpVerificationCode = getTotpCode(
+      totpSecret.secretKey,
+      totpSecret.codeIntervalSeconds,
+      totpSecret.codeLength
+    );
+
+    const multiFactorAssertion =
+      TotpMultiFactorGenerator.assertionForEnrollment(
+        totpSecret,
+        totpVerificationCode
+      );
+    await expect(mfaUser.enroll(multiFactorAssertion, displayName)).to.be
+      .fulfilled;
+  });
+
+  it('should not allow sign-in with incorrect totp', async () => {
+    let resolver;
+    // Added a delay so that getTotpCode() actually generates a new totp code
+    await delay(30 * 1000);
```

Comment:
These 30 second delays are going to cause a much longer presubmit :(

We should investigate getting these tests into the emulator ASAP

Reply:
I don't think a timeout is necessary here.. we can just pass in a known invalid code like "abcde" for this test case.. valid codes will be numbers.